In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.decomposition import PCA
from sklearn.svm import SVR
from sklearn.metrics import r2_score, mean_squared_error
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('07_flats_EDA.csv')
df.dropna(inplace=True)

In [3]:
df['price_cr'] = df['price_cr']*1.7    #since, the dataset is 4 years old, increasing prices by 70%

In [4]:
df.head(2)

,price_cr,area_sqm,bhk,bathroom,balcony,addl_room,sector,landmark_count,has_hospital_or_clinic,has_police_stn,has_fire_stn,has_mall,has_market,has_park,has_bank_or_atm,has_institute,has_airport,has_rly_stn,has_metro_stn,has_expy,has_petrol_pump,has_hotel,has_temple_or_church,has_famous_chowk,has_club,has_stadium,has_restro_cafe,area_type,floor_rise,age,society,city
0,0.765,83.61,2,2,1,0,7,50,34,0,0,0,0,0,5,2,0,0,0,0,0,0,1,0,0,0,1,carpet,mid,new,maa bhagwati residency,gurgaon
1,0.850,60.39,2,2,1,0,3,48,4,0,0,0,0,0,1,0,0,0,0,0,0,0,2,0,0,0,0,carpet,low,old,apna enclave,gurgaon


In [5]:
le = LabelEncoder()
for col in df.columns:
    if df[col].dtype == 'object':
        df[col] = le.fit_transform(df[col])

In [6]:
X = df.iloc[:,1:]
y = df['price_cr']

In [7]:
#train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
#Scaling X
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

In [9]:
X_train_scaled.shape[1]

31

In [10]:
X_train_scaled = pd.DataFrame(X_train_scaled)

In [11]:
#using PCA to reduce dimensionality
pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

In [12]:
X_train_pca.shape[1]

25

In [13]:
pca.components_

array([[-7.25455215e-02, -5.35575528e-02, -9.21091740e-02,
        -1.12081381e-01, -1.07964809e-01, -2.20018060e-01,
         3.47693899e-01,  2.42324693e-01,  2.26626437e-02,
         1.15863277e-02, -2.45103484e-01,  1.59659950e-01,
        -2.09746092e-02,  3.49120181e-01, -1.27700406e-01,
        -2.78677860e-01,  1.97977359e-03,  1.75167752e-01,
        -1.83780957e-01,  1.50463933e-01, -1.90489624e-01,
         3.35682322e-01,  1.13371487e-01,  8.63939763e-03,
        -1.06589667e-01,  3.14509483e-01, -1.51555919e-01,
         9.39513419e-02,  1.90522508e-01,  7.16968509e-04,
         3.55696884e-02],
       [ 4.04244866e-01,  4.17145561e-01,  4.33976009e-01,
         3.09640687e-01,  3.00736358e-01,  4.61098664e-02,
         1.48074029e-01,  2.95608541e-02,  1.62854766e-02,
         1.17035340e-02,  4.02298296e-02,  1.14694946e-01,
         1.69284064e-01,  1.86303551e-01,  8.77837001e-02,
         6.55415756e-02, -3.52841166e-02,  2.05781195e-01,
        -3.12844728e-02, -4.82

In [14]:
pca.explained_variance_

array([5.2472799 , 3.45937659, 2.01442137, 1.52980141, 1.39579944,
       1.29147144, 1.23355386, 1.13662271, 1.11453307, 1.02852424,
       0.9581142 , 0.92733915, 0.9204982 , 0.82622318, 0.77583628,
       0.74726284, 0.70811883, 0.67120174, 0.65628441, 0.61667396,
       0.57291401, 0.53955047, 0.50498324, 0.42265459, 0.37258103])

In [15]:
pca.explained_variance_ratio_

array([0.16919423, 0.11154475, 0.06495336, 0.04932719, 0.04500641,
       0.04164244, 0.03977493, 0.03664947, 0.0359372 , 0.03316392,
       0.0308936 , 0.02990129, 0.02968071, 0.02664089, 0.0250162 ,
       0.02409488, 0.02283271, 0.02164235, 0.02116135, 0.01988414,
       0.01847314, 0.01739736, 0.01628277, 0.01362815, 0.01201357])

In [16]:
pca.explained_variance_ratio_.sum()

0.9567370054241593

In [17]:
y_train.unique()

array([ 0.765,  5.695,  1.87 ,  1.53 ,  2.856,  3.145,  4.165,  0.799,
        2.499,  2.278,  0.901,  2.312,  3.06 ,  5.95 ,  2.55 ,  2.244,
        1.972,  5.083,  1.836,  0.425,  2.38 ,  1.683,  0.697, 14.025,
        0.85 ,  2.635,  4.93 ,  2.125, 11.05 ,  1.02 ,  8.5  ,  1.785,
        3.247,  0.918,  4.454,  0.51 ,  1.36 ,  0.374,  0.646,  2.72 ,
        2.465,  1.479,  1.7  ,  1.904,  2.04 ,  2.89 ,  4.76 ,  0.714,
        6.97 ,  2.295,  2.414,  8.075,  0.68 ,  2.975,  2.992,  1.275,
        0.816,  2.006,  3.23 ,  5.015,  1.955,  6.29 ,  2.941,  1.666,
        6.8  ,  3.111,  8.398,  1.496,  5.593,  2.21 ,  3.179,  6.205,
        1.615,  0.476,  3.094, 13.26 ,  3.91 ,  0.442,  2.363,  2.227,
        2.822,  3.485,  1.445,  3.689,  6.511,  1.513,  2.142,  2.533,
        3.451,  0.663,  1.19 ,  3.995, 15.3  ,  6.12 ,  0.782,  2.805,
        3.026,  4.505,  0.459,  1.326,  1.428,  0.561,  1.632,  1.377,
        4.267,  3.4  ,  5.423,  4.59 ,  1.411,  0.595,  4.08 ,  4.845,
      

In [18]:
#model 1: RF Regression
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 1.1002795327346713
R2_score : 0.755566915850097


In [19]:
#model 2: XG Boost Regression
from xgboost import XGBRegressor
model = XGBRegressor(objective='reg:squarederror',n_estimators=100, 
                         learning_rate=0.1,max_depth=5,random_state=42)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 1.0318210541924913
R2_score : 0.7850375072748078


In [20]:
#model 3: SVR Regression
from sklearn.svm import SVR
model = SVR(kernel='rbf', C=100, gamma=0.1, epsilon=.1)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 0.906435709425613
R2_score : 0.8341070493889371


In [21]:
#model 4: SVR Regression
from sklearn.svm import SVR
model = SVR(kernel='rbf', C=70, gamma=0.06, epsilon=.14)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 0.8324428875710824
R2_score : 0.8600854716169202


#### the best model is Support Vector Regressor (SVR) with 86% R2 Score